selfservice/strategy/code/strategy.go

```diff
@@ -82,14 +118,319 @@ func NewStrategy(deps strategyDependencies) *Strategy {
+			}
+
+			for _, n := range traitNodes {
+				nodes.Append(n)
+			}
+		}
+
+		var codeMetaLabel *text.Message
+
+		switch f.GetFlowName() {
+		case flow.VerificationFlow, flow.RecoveryFlow:
+			codeMetaLabel = text.NewInfoNodeLabelSubmit()
+		case flow.LoginFlow:
+			codeMetaLabel = text.NewInfoSelfServiceLoginCode()
+		case flow.RegistrationFlow:
+			codeMetaLabel = text.NewInfoSelfServiceRegistrationRegisterCode()
+		}
+
+		methodButton := node.NewInputField("method", s.ID(), node.CodeGroup, node.InputAttributeTypeSubmit).
+			WithMetaLabel(codeMetaLabel)
+
+		nodes.Append(methodButton)
+
+		f.GetUI().Nodes = nodes
+
+	case flow.StateEmailSent:
+		// fresh ui node group
+		freshNodes := node.Nodes{}
+		var route string
+		var codeMetaLabel *text.Message
+		var message *text.Message
+
+		var resendNode *node.Node
+
+		switch f.GetFlowName() {
+		case flow.RecoveryFlow:
+			route = recovery.RouteSubmitFlow
+			codeMetaLabel = text.NewInfoNodeLabelRecoveryCode()
+			message = text.NewRecoveryEmailWithCodeSent()
+
+			resendNode = node.NewInputField("email", nil, node.CodeGroup, node.InputAttributeTypeEmail, node.WithRequiredInputAttribute).
+				WithMetaLabel(text.NewInfoNodeResendOTP())
+		case flow.VerificationFlow:
+			route = verification.RouteSubmitFlow
+			codeMetaLabel = text.NewInfoNodeLabelVerificationCode()
+			message = text.NewVerificationEmailWithCodeSent()
+
+		case flow.LoginFlow:
+			route = login.RouteSubmitFlow
+			codeMetaLabel = text.NewInfoNodeLabelLoginCode()
+			message = text.NewLoginEmailWithCodeSent()
+
+			// preserve the login identifier that were submitted
+			// so we can retry the code flow with the same data
+			for _, n := range f.GetUI().Nodes {
+				if n.Group == node.DefaultGroup {
+					freshNodes = append(freshNodes, n)
+				}
+			}
+
+			resendNode = node.NewInputField("resend", "code", node.CodeGroup, node.InputAttributeTypeSubmit).
+				WithMetaLabel(text.NewInfoNodeResendOTP())
+
+		case flow.RegistrationFlow:
+			route = registration.RouteSubmitFlow
+			codeMetaLabel = text.NewInfoNodeLabelRegistrationCode()
+			message = text.NewRegistrationEmailWithCodeSent()
+
+			// in the registration flow we need to preserve the trait fields that were submitted
+			// so we can retry the code flow with the same data
+			for _, n := range f.GetUI().Nodes {
+				if t, ok := n.Attributes.(*node.InputAttributes); ok && t.Type == node.InputAttributeTypeSubmit {
+					continue
+				}
+
+				if n.Group == node.CodeGroup {
+					freshNodes = append(freshNodes, n)
+				}
+			}
+
+			resendNode = node.NewInputField("resend", "code", node.CodeGroup, node.InputAttributeTypeSubmit).
+				WithMetaLabel(text.NewInfoNodeResendOTP())
+		default:
+			return errors.New("received unexpected flow type")
+		}
+
+		// Hidden field Required for the re-send code button
+		// !!important!!: this field must be appended before the code submit button since upsert will replace the first node with the same name
+		freshNodes.Upsert(
```

Comment:
I'm not sure if I get this logic?

1. We're adding `<input type="hidden" name="method" value="code">`
2. We're adding `<input type="text" required name="code">`
3. We're changing input from 1 to `<input type="submit" name="method" value="code">`

Reply:
No, the upsert overwrites any input's with the name `method`. So we need to add it to the UI nodes first, but we use upsert here since you could be re-submitting the flow and it could be throwing an error, such as traits do not match or the email is required etc. 

So this upsert will replace the current `<input type="hidden" name="method" value="code">` and then we append the submit button again since it was removed earlier. 

Between submits we essentially always remove the submit types and re-add them. The way we do `Upserts` within nodes are thus where the problem is since it's not a true check of similarity. So this is a work-around to that.